worker manager panic because creating a duplicated worker(job master actually)

Steps to reproduce

1. Start three server masters, one executor
2. Submit one demo job
3. Restart the executor server
4. After a while, the leader server master will failover demo job(create new job master via CreateWorker) and panic

```go
[2022/04/24 18:29:32.145 +08:00] [INFO] [jobmanager.go:308] ["on worker offline"] [id=503bcd85-7d77-4bc5-8a5c-71764e6fe206] [reason="[DFLOW:ErrWorkerOffline]worker is offline: workerID: 503bcd85-7d77-4bc5-8a5c-71764e6fe206, error message: %!s(MISSING)"]
[2022/04/24 18:29:32.145 +08:00] [INFO] [master.go:489] [CreateWorker] [worker-type=3] [worker-config="{\"SeqID\":0,\"CreatedAt\":\"0001-01-01T00:00:00Z\",\"UpdatedAt\":\"0001-01-01T00:00:00Z\",\"project-id\":\"\",\"id\":\"503bcd85-7d77-4bc5-8a5c-71764e6fe206\",\"type\":3,\"status\":1,\"node-id\":\"\",\"addr\":\"\",\"epoch\":0,\"config\":\"ewogICAgImpvYi1uYW1lIjogInRlc3QtZmFrZS1qb2ItMSIsCiAgICAid29ya2VyLWNvdW50IjogMywKICAgICJ0YXJnZXQtdGljayI6IDIwMAp9Cg==\"}"] [master-id=dataflow-engine-job-manager] 
[2022/04/24 18:29:32.145 +08:00] [INFO] [job_fsm.go:189] ["job master recovered"] [job="{\"SeqID\":0,\"CreatedAt\":\"0001-01-01T00:00:00Z\",\"UpdatedAt\":\"0001-01-01T00:00:00Z\",\"project-id\":\"\",\"id\":\"503bcd85-7d77-4bc5-8a5c-71764e6fe206\",\"type\":3,\"status\":1,\"node-id\":\"\",\"addr\":\"\",\"epoch\":0,\"config\":\"ewogICAgImpvYi1uYW1lIjogInRlc3QtZmFrZS1qb2ItMSIsCiAgICAid29ya2VyLWNvdW50IjogMywKICAgICJ0YXJnZXQtdGljayI6IDIwMAp9Cg==\"}"]
[2022/04/24 18:29:32.146 +08:00] [PANIC] [worker_manager.go:317] ["worker already exists"] [worker-id=503bcd85-7d77-4bc5-8a5c-71764e6fe206] [stack="github.com/hanfei1991/microcosm/lib/master.(*WorkerManager).OnCreatingWorker\n\t/home/apple/work/code/microcosm/lib/master/worker_manager.go:317\ngithub.com/hanfei1991/microcosm/lib.(*DefaultBaseMaster).CreateWorker.func1\n\t/home/apple/work/code/microcosm/lib/master.go:535"]
panic: worker already exists

goroutine 1692 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc00018ecc0, {0xc0025c3b00, 0x1, 0x1})
        /home/apple/.gvm/pkgsets/go1.18/global/pkg/mod/go.uber.org/zap@v1.21.0/zapcore/entry.go:232 +0x44c
go.uber.org/zap.(*Logger).Panic(0x1b03620?, {0x1e3ebd5?, 0xc0002e76e0?}, {0xc0025c3b00, 0x1, 0x1})
        /home/apple/.gvm/pkgsets/go1.18/global/pkg/mod/go.uber.org/zap@v1.21.0/logger.go:230 +0x59
github.com/hanfei1991/microcosm/lib/master.(*WorkerManager).OnCreatingWorker(0xc000a1edd0, {0xc0002e76e0, 0x24}, {0xc0025957a0, 0x24})
        /home/apple/work/code/microcosm/lib/master/worker_manager.go:317 +0x1fa
github.com/hanfei1991/microcosm/lib.(*DefaultBaseMaster).CreateWorker.func1()
        /home/apple/work/code/microcosm/lib/master.go:535 +0x6bd
created by github.com/hanfei1991/microcosm/lib.(*DefaultBaseMaster).CreateWorker
        /home/apple/work/code/microcosm/lib/master.go:506 +0x638
```

---

- job master has an unchanged name after it is created.
- worker manager doesn't remove offline worker entry currently.

--- a/lib/master/worker_handle.go
+++ b/lib/master/worker_handle.go
@@ -32,10 +32,16 @@ type RunningHandle interface {
 	) error
 }
 
+// TombstoneHandle represents a dead worker.
 type TombstoneHandle interface {
 	Status() *libModel.WorkerStatus
 	ID() libModel.WorkerID
 	ToPB() (*pb.WorkerInfo, error)
+
+	// CleanTombstone cleans the metadata from the metastore,
+	// and cleans the state managed by the framework.
+	// Do not call any other methods on this handle after
+	// CleanTombstone is called.
 	CleanTombstone(ctx context.Context) error
 }
 
@@ -139,6 +145,11 @@ func (h *tombstoneHandleImpl) CleanTombstone(ctx context.Context) error {
 	}
 	if !ok {
 		log.L().Info("Tombstone already cleaned", zap.String("worker-id", h.workerID))
+		// Idempotent for robustness.
+		return nil
 	}
+	log.L().Info("Worker tombstone is cleaned", zap.String("worker-id", h.workerID))
+	h.manager.removeTombstoneEntry(h.workerID)
+
 	return nil
 }

--- a/lib/master/worker_manager.go
+++ b/lib/master/worker_manager.go
@@ -555,3 +555,24 @@ func (m *WorkerManager) enqueueEvent(event *masterEvent) error {
 
 	return nil
 }
+
+// removeTombstoneEntry removes a tombstone workerEntry from the in-memory map.
+// NOTE: removeTombstoneEntry is expected to be used by tombstoneHandleImpl only,
+// and it should NOT be called with m.mu taken.
+func (m *WorkerManager) removeTombstoneEntry(id libModel.WorkerID) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Checks precondition.
+	entry, exists := m.workerEntries[id]
+	if !exists {
+		// Return here. We intend this method to be idempotent.
+		return
+	}
+
+	if !entry.IsTombstone() {
+		log.L().Panic("Unreachable: not a tombstone", zap.Stringer("entry", entry))
+	}
+
+	delete(m.workerEntries, id)
+}

--- a/lib/master/worker_manager_test.go
+++ b/lib/master/worker_manager_test.go
@@ -12,6 +12,7 @@ import (
 	"golang.org/x/time/rate"
 
 	"github.com/hanfei1991/microcosm/lib/config"
+	"github.com/hanfei1991/microcosm/lib/metadata"
 	libModel "github.com/hanfei1991/microcosm/lib/model"
 	"github.com/hanfei1991/microcosm/lib/statusutil"
 	"github.com/hanfei1991/microcosm/pkg/adapter"
@@ -385,6 +386,37 @@ func TestRecoverWithNoWorker(t *testing.T) {
 	// InitAfterRecover returning at all would indicate a successful test.
 	err := suite.manager.InitAfterRecover(ctx)
 	require.NoError(t, err)
+
+	suite.Close()
+}
+
+func TestCleanTombstone(t *testing.T) {
+	ctx := context.Background()
+
+	suite := NewWorkerManageTestSuite(true)
+	suite.manager.OnCreatingWorker("worker-1", "executor-1")
+	suite.AdvanceClockBy(30 * time.Second)
+	suite.AdvanceClockBy(30 * time.Second)
+	suite.AdvanceClockBy(30 * time.Second)
+
+	event := suite.WaitForEvent(t, "worker-1")
+	require.Equal(t, workerOfflineEvent, event.Tp)
+	require.NotNil(t, event.Handle.GetTombstone())
+	err := event.Handle.GetTombstone().CleanTombstone(ctx)
+	require.NoError(t, err)
+
+	workerMetaClient := metadata.NewWorkerMetadataClient("master-1", suite.meta)
+	_, err = workerMetaClient.Load(ctx, "worker-1")
+	// Asserts that the meta for the worker is indeed deleted.
+	require.Error(t, err)
+	require.Regexp(t, ".*ErrWorkerNoMeta", err)
+
+	// CleanTombstone should be idempotent for robustness.
+	err = event.Handle.GetTombstone().CleanTombstone(ctx)
+	require.NoError(t, err)
+
+	// Recreating a worker with the same name should work fine.
+	suite.manager.OnCreatingWorker("worker-1", "executor-1")
 
 	suite.Close()
 }

